src/ipc/notification_pool.c

```diff
@@ -0,0 +1,81 @@
+// SPDX-License-Identifier: BSD-3-Clause
+//
+// Copyright(c) 2025 Intel Corporation. All rights reserved.
+//
+// Author: Adrian Warecki <adrian.warecki@intel.com>
+//
+
+#include <stdint.h>
+#include <sof/common.h>
+#include <sof/list.h>
+#include <sof/ipc/notification_pool.h>
+
+#define NOTIFICATION_POOL_MAX_PAYLOAD_SIZE	40	/* IPC4 Resource Event needs 10dw */
+
+LOG_MODULE_REGISTER(notification_pool, CONFIG_SOF_LOG_LEVEL);
+
+SOF_DEFINE_REG_UUID(notification_pool);
+
+DECLARE_TR_CTX(notif_tr, SOF_UUID(notification_pool_uuid), LOG_LEVEL_INFO);
+
+struct ipc_notif_pool_item {
+	struct ipc_msg msg;
+	uint32_t payload[SOF_DIV_ROUND_UP(NOTIFICATION_POOL_MAX_PAYLOAD_SIZE, sizeof(uint32_t))];
+};
+
+static struct list_item pool_free_list = LIST_INIT(pool_free_list);
+struct k_spinlock pool_free_list_lock;
+
+static void ipc_notif_free(struct ipc_msg *msg)
+{
+	struct ipc_notif_pool_item *item = container_of(msg, struct ipc_notif_pool_item, msg);
+	k_spinlock_key_t key;
+
+	key = k_spin_lock(&pool_free_list_lock);
+	list_item_append(&item->msg.list, &pool_free_list);
+	k_spin_unlock(&pool_free_list_lock, key);
+}
+
+static struct ipc_msg *ipc_notif_new(size_t size)
+{
+	struct ipc_notif_pool_item *item;
+
+	item = rzalloc(SOF_MEM_ZONE_RUNTIME_SHARED, 0, SOF_MEM_CAPS_RAM, sizeof(*item));
+	if (!item) {
+		tr_err(&notif_tr, "Unable to allocate memory for notification message");
+		return NULL;
+	}
+
+	list_init(&item->msg.list);
+	item->msg.tx_data = item->payload;
+	item->msg.tx_size = size;
+	item->msg.callback = ipc_notif_free;
+	return &item->msg;
+}
+
+struct ipc_msg *ipc_notification_pool_get(size_t size)
+{
+	struct ipc_notif_pool_item *item;
+	k_spinlock_key_t key;
+
+	if (size > NOTIFICATION_POOL_MAX_PAYLOAD_SIZE) {
+		tr_err(&notif_tr, "Requested size %zu exceeds maximum payload size %u",
+		       size, NOTIFICATION_POOL_MAX_PAYLOAD_SIZE);
+		return NULL;
+	}
+
+	/* check if we have a free message */
+	key = k_spin_lock(&pool_free_list_lock);
+	if (list_is_empty(&pool_free_list)) {
+		k_spin_unlock(&pool_free_list_lock, key);
+		return ipc_notif_new(size);
```

Comment:
Should there be some sanity check on how many items can be at max allocated? E.g. if there's a problem/delay sending the notification out, LL scheduler could flood this interface with pool_get()s and exhaust memory.

Reply:
I agree, will add it.